File: app/src/main/java/com/jeketos/associatewith/guesser/GuesserPresenterImpl.java
package com.jeketos.associatewith.guesser;

import com.google.firebase.database.DataSnapshot;
import com.jeketos.associatewith.Point;
import com.jeketos.associatewith.di.Injector;
import com.jeketos.associatewith.guesser.chat.ChatItem;
import com.jeketos.associatewith.guesser.chat.IChatItem;

import java.util.HashMap;

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */

public class GuesserPresenterImpl implements GuesserMVP.GuesserPresenter {

    private GuesserMVP.GuesserView view;
    private GuesserMVP.GuesserModel model;
    private float previousX, previousY;
    private int previousPointCount;
    private int previousChatCount;

    public GuesserPresenterImpl(GuesserMVP.GuesserView view) {
        this.view = view;
        previousX = 0;
        previousY = 0;
        previousPointCount = 0;
        model = Injector.provideGuesserModel(this);
        view.init();
    }

    @Override
    public void dataReceived(DataSnapshot dataSnapshot) {
        int childrenCount = ((Number) dataSnapshot.getChildrenCount()).intValue();
        if(dataSnapshot.getChildrenCount() == 0){
            view.clearBoard();
        } else {
            for (int i = previousPointCount; i < childrenCount; i++) {
                int motionEvent = ((Number) ((HashMap) dataSnapshot.child(Integer.toString(i)).getValue()).get("motionEvent")).intValue();
                float x = getFloatValue(((HashMap) dataSnapshot.child(Integer.toString(i)).getValue()).get("x"));
                float y = getFloatValue(((HashMap) dataSnapshot.child(Integer.toString(i)).getValue()).get("y"));
                Point point = new Point(x, y, motionEvent);
                view.draw(previousX, previousY, point);
                previousX = x;
                previousY = y;
            }
        }
        previousPointCount = childrenCount;

    }

    @Override
    public void chatDataReceived(DataSnapshot dataSnapshot) {
        int childrenCount = ((Number) dataSnapshot.getChildrenCount()).intValue();
        if(dataSnapshot.getChildrenCount() == 0){
            view.clearChat();
        } else {
            for (int i = previousChatCount; i < childrenCount; i++) {
                String name = (String) ((HashMap) dataSnapshot.child(Integer.toString(i)).getValue()).get(ChatItem.NAME);
                String message = (String) ((HashMap) dataSnapshot.child(Integer.toString(i)).getValue()).get(ChatItem.MESSAGE);
                IChatItem item = new ChatItem();
                item.setName(name);
                item.setMessage(message);
                view.addChatItem(item);
            }
        }
        previousChatCount = childrenCount;
    }

    @Override
    public void sendMessage(String message) {
        IChatItem item = new ChatItem();
        item.setName(android.os.Build.MODEL);
        item.setMessage(message);
        model.sendMessage(item);
    }

    private float getFloatValue(Object value){
        return ((Number) value).floatValue();
    }

}

File: app/src/main/java/com/jeketos/associatewith/guesser/GuesserPresenterImpl.kt
package com.jeketos.associatewith.guesser

import android.text.TextUtils
import com.google.firebase.database.DataSnapshot
import com.jeketos.associatewith.Point
import com.jeketos.associatewith.base.BaseMvpPresenter
import com.jeketos.associatewith.guesser.chat.ChatItem
import com.jeketos.associatewith.util.ChatUtils
import java.util.*
import javax.inject.Inject

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */

class GuesserPresenterImpl @Inject constructor() : BaseMvpPresenter<GuesserMVP.GuesserView>(), GuesserMVP.GuesserPresenter {


     @Inject lateinit var model : GuesserMVP.GuesserModel
     var previousX : Float = 0f
     var previousY : Float = 0f
     var previousPointCount : Int = 0
     var previousChatCount : Int = 0
     var selectedWord : String? = null

    override fun init() {
        previousX = 0f
        previousY = 0f
        previousPointCount = 0
    }

    override fun moveDataReceived(dataSnapshot: DataSnapshot) {
        val childrenCount = dataSnapshot.childrenCount.toInt()
        if(childrenCount == 0){
            view!!.clearBoard()
        } else {
            for (i in previousPointCount until childrenCount){
                val hashMap = dataSnapshot.child(i.toString()).value as HashMap<*,*>
                val motionEvent = hashMap["motionEvent"] as Long
                val x = hashMap["x"].toString().toFloat()
                val y = hashMap["y"].toString().toFloat()
                view!!.draw(previousX, previousY, Point(x,y,motionEvent.toInt()))
                previousX = x
                previousY = y
            }
        }
        previousPointCount = childrenCount
    }

    override fun chatDataReceived(dataSnapshot: DataSnapshot) {
        val childrenCount = dataSnapshot.childrenCount.toInt()
        if(childrenCount == 0){
            view!!.clearChat()
        } else {
            val list = Array(childrenCount - previousChatCount, { i -> i + previousChatCount})
            list.forEach {
                view!!.addChatItem(ChatUtils.getChatItem(dataSnapshot, it))
            }
            previousChatCount = childrenCount
        }
    }

    override fun selectedWordReceived(dataSnapshot: DataSnapshot) {
        selectedWord = dataSnapshot.value as? String
    }

    override fun winnerDataReceived(dataSnapshot: DataSnapshot) {
        if(dataSnapshot.value != null) {
            val name = dataSnapshot.value as String
            if (!TextUtils.isEmpty(name)) {
                view!!.showWinnerDialog(name, selectedWord!!)
            }
        }
    }

    override fun sendMessage(message: String) {
        val item = ChatItem()
        item.setName(android.os.Build.MODEL)
        item.setMessage(message)
        model.sendMessage(item)
        if(message.toLowerCase() == (selectedWord)){
            model.setWinner(item)
            view!!.showWinnerDialog(item.getName(), message)
        }
    }

    override fun attachView(view: GuesserMVP.GuesserView) {
        super.attachView(view)
        model.addEventListeners()
        model.addChatListener { dataSnapshot: DataSnapshot -> chatDataReceived(dataSnapshot) }
        model.addMoveListener { dataSnapshot: DataSnapshot -> moveDataReceived(dataSnapshot) }
        model.addWinnerListener { dataSnapshot: DataSnapshot ->  winnerDataReceived(dataSnapshot)}
        model.addSelectedWordListener { dataSnapshot: DataSnapshot ->  selectedWordReceived(dataSnapshot)}
    }

    override fun detachView() {
        super.detachView()
        model.removeListeners()
    }
}
